diff --git a/src/main/scala/coupledL2/CoupledL2.scala b/src/main/scala/coupledL2/CoupledL2.scala
index a2e0487..d4f77f3 100644
--- a/src/main/scala/coupledL2/CoupledL2.scala
+++ b/src/main/scala/coupledL2/CoupledL2.scala
@@ -364,7 +364,7 @@ abstract class CoupledL2Base(implicit p: Parameters) extends LazyModule with Has
       case EdgeOutKey => node.out.head._2
       case BankBitsKey => bankBits
     }
-    val prefetcher = prefetchOpt.map(_ => Module(new Prefetcher()(pftParams)))
+    lazy val prefetcher = prefetchOpt.map(_ => Module(new Prefetcher()(pftParams)))
     val prefetchTrains = prefetchOpt.map(_ => Wire(Vec(banks, DecoupledIO(new PrefetchTrain()(pftParams)))))
     val prefetchResps = prefetchOpt.map(_ => Wire(Vec(banks, DecoupledIO(new PrefetchResp()(pftParams)))))
     val prefetchReqsReady = WireInit(VecInit(Seq.fill(banks)(false.B)))
@@ -414,26 +414,30 @@ abstract class CoupledL2Base(implicit p: Parameters) extends LazyModule with Has
     val releaseSourceD = Wire(Vec(banks, Bool()))
     val allCanFire = (RegNextN(!hintFire, sliceAhead) && RegNextN(!hintFire, sliceAhead + 1)) || Cat(releaseSourceD).orR
 
+    def createSlice(edgeIn: TLEdgeIn, edgeOut: TLEdgeOut, i: Int): BaseSlice[_ <: BaseOuterBundle] = {
+      if (enableCHI) {
+        Module(new tl2chi.Slice()(p.alterPartial {
+          case EdgeInKey => edgeIn
+          case EdgeOutKey => edgeOut
+          case BankBitsKey => bankBits
+          case SliceIdKey => i
+        }))
+      } else {
+        Module(new tl2tl.Slice()(p.alterPartial {
+          case EdgeInKey => edgeIn
+          case EdgeOutKey => edgeOut
+          case BankBitsKey => bankBits
+          case SliceIdKey => i
+        }))
+      }
+    }
+
     val slices = node.in.zip(node.out).zipWithIndex.map {
       case (((in, edgeIn), (out, edgeOut)), i) =>
         require(in.params.dataBits == out.params.dataBits)
         val rst_L2 = reset
         val slice = withReset(rst_L2) {
-          if (enableCHI) {
-            Module(new tl2chi.Slice()(p.alterPartial {
-              case EdgeInKey => edgeIn
-              case EdgeOutKey => edgeOut
-              case BankBitsKey => bankBits
-              case SliceIdKey => i
-            }))
-          } else {
-            Module(new tl2tl.Slice()(p.alterPartial {
-              case EdgeInKey => edgeIn
-              case EdgeOutKey => edgeOut
-              case BankBitsKey => bankBits
-              case SliceIdKey => i
-            }))
-          }
+          createSlice(edgeIn, edgeOut, i)
         }
         slice.io.in <> in
         if (enableHintGuidedGrant) {
diff --git a/src/main/scala/coupledL2/Directory.scala b/src/main/scala/coupledL2/Directory.scala
index 67003d2..b846b74 100644
--- a/src/main/scala/coupledL2/Directory.scala
+++ b/src/main/scala/coupledL2/Directory.scala
@@ -138,37 +138,37 @@ class Directory(implicit p: Parameters) extends L2Module {
   val metaWen = io.metaWReq.valid
   val replacerWen = WireInit(false.B)
 
-  // val tagArray  = Module(new SRAMTemplate(UInt(tagBits.W), sets, ways, singlePort = true))
+  val tagArray  = Module(new SRAMTemplate(UInt(tagBits.W), sets, ways, singlePort = true))
   private val mbist = p(L2ParamKey).hasMbist
   private val hasSramCtl = p(L2ParamKey).hasSramCtl
-  val tagArray = if (enableTagECC) {
-    Module(new SplittedSRAM(
-      gen = UInt((tagBankSplit * encTagBankBits).W),
-      set = sets,
-      way = ways,
-      waySplit = 2,
-      dataSplit = if (enableTagSRAMSplit) {
-        tagSRAMSplit
-      } else {
-        1
-      },
-      singlePort = true,
-      readMCP2 = false,
-      hasMbist = mbist,
-      hasSramCtl = hasSramCtl
-    ))
-  } else {
-    Module(new SplittedSRAM(
-      gen = UInt(tagBits.W),
-      set = sets,
-      way = ways,
-      waySplit = 2,
-      singlePort = true,
-      readMCP2 = false,
-      hasMbist = mbist,
-      hasSramCtl = hasSramCtl
-    ))
-  }
+  // val tagArray = if (enableTagECC) {
+  //   Module(new SplittedSRAM(
+  //     gen = UInt((tagBankSplit * encTagBankBits).W),
+  //     set = sets,
+  //     way = ways,
+  //     waySplit = 2,
+  //     dataSplit = if (enableTagSRAMSplit) {
+  //       tagSRAMSplit
+  //     } else {
+  //       1
+  //     },
+  //     singlePort = true,
+  //     readMCP2 = false,
+  //     hasMbist = mbist,
+  //     hasSramCtl = hasSramCtl
+  //   ))
+  // } else {
+  //   Module(new SplittedSRAM(
+  //     gen = UInt(tagBits.W),
+  //     set = sets,
+  //     way = ways,
+  //     waySplit = 2,
+  //     singlePort = true,
+  //     readMCP2 = false,
+  //     hasMbist = mbist,
+  //     hasSramCtl = hasSramCtl
+  //   ))
+  // }
 
   val metaArray = Module(new SRAMTemplate(new MetaEntry, sets, ways, singlePort = true, hasMbist = mbist, hasSramCtl = hasSramCtl))
 
diff --git a/src/main/scala/coupledL2/tl2tl/MainPipe.scala b/src/main/scala/coupledL2/tl2tl/MainPipe.scala
index 605ca02..9d6636a 100644
--- a/src/main/scala/coupledL2/tl2tl/MainPipe.scala
+++ b/src/main/scala/coupledL2/tl2tl/MainPipe.scala
@@ -599,8 +599,8 @@ class MainPipe(implicit p: Parameters) extends L2Module with HasPerfEvents {
   io.toReqArb.blockB_s1 :=
     task_s2.valid && bBlock(task_s2.bits) ||
     task_s3.valid && bBlock(task_s3.bits) ||
-    task_s4.valid && bBlock(task_s4.bits, tag = true) ||
-    task_s5.valid && bBlock(task_s5.bits, tag = true)
+    task_s4.valid && bBlock(task_s4.bits, tag = true) && task_s4.bits.opcode(2, 1) === Grant(2, 1) ||
+    task_s5.valid && bBlock(task_s5.bits, tag = true) && task_s5.bits.opcode(2, 1) === Grant(2, 1)
 
   io.toReqArb.blockA_s1 := false.B
 
